components/merkledb/src/proof_entry.rs

```diff
@@ -0,0 +1,296 @@
+// Copyright 2019 The Exonum Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! An implementation of index that may only contain one element.
+
+use std::marker::PhantomData;
+
+use exonum_crypto::Hash;
+
+use crate::{
+    access::{Access, AccessError, FromAccess},
+    views::{IndexAddress, IndexState, IndexType, RawAccess, RawAccessMut, View, ViewWithMetadata},
+    BinaryValue, ObjectHash,
+};
+
+/// A Merkelized index that may only contain one element.
+///
+/// You can add an element to this index and check whether it exists. A value
+/// should implement [`BinaryValue`] and [`ObjectHash`] traits. Unlike [`Entry`],
+/// hashed entries are eagerly hashed and may participate in [state aggregation].
+///
+/// [`BinaryValue`]: trait.BinaryValue.html
+/// [`ObjectHash`]: trait.ObjectHash.html
+/// [`Entry`]: struct.Entry.html
+/// [state aggregation]: index.html#state-aggregation
+#[derive(Debug)]
+pub struct ProofEntry<T: RawAccess, V> {
```

Comment:
It could be possible to customize `Entry` instead:

```rust
struct Entry<T: RawAccess, V, Type: EntryType> { /* ...*/ }
```

...in order to minimize the amount of copy-pasted code.